src/qengine/state/opencl.cpp

```diff
@@ -10,47 +10,59 @@
 // See LICENSE.md in the project root or https://www.gnu.org/licenses/gpl-3.0.en.html
 // for details.
 
+#include <memory>
+
 #include "oclengine.hpp"
 #include "qengine_opencl.hpp"
 
 namespace Qrack {
 
 #define CMPLX_NORM_LEN 5
 
-void QEngineOCL::InitOCL()
+void QEngineOCL::SetDevice(const int& dID)
 {
-    clObj = OCLEngine::Instance();
+    if (dID >= 0) {
+        deviceID = dID % (clObj->GetDeviceCount());
+    } else {
+        deviceID = -1;
+    }
+    device_context = clObj->GetDeviceContextPtr(deviceID);
+    context = device_context->context;
+    queue = device_context->queue;
+}
 
-    queue = *(clObj->GetQueuePtr());
-    cl::Context context = *(clObj->GetContextPtr());
+void QEngineOCL::InitOCL(int devID)
+{
+    clObj = OCLEngine::Instance();
+    SetDevice(devID);
 
     // create buffers on device (allocate space on GPU)
-    stateBuffer = cl::Buffer(context, CL_MEM_USE_HOST_PTR | CL_MEM_READ_WRITE, sizeof(complex) * maxQPower, stateVec);
+    stateBuffer = std::make_shared<cl::Buffer>(
+        context, CL_MEM_USE_HOST_PTR | CL_MEM_READ_WRITE, sizeof(complex) * maxQPower, stateVec);
     cmplxBuffer = cl::Buffer(context, CL_MEM_READ_ONLY, sizeof(complex) * 5);
     ulongBuffer = cl::Buffer(context, CL_MEM_READ_ONLY, sizeof(bitCapInt) * 10);
     nrmBuffer = cl::Buffer(context, CL_MEM_ALLOC_HOST_PTR | CL_MEM_READ_WRITE,
         sizeof(real1) * CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE);
     maxBuffer = cl::Buffer(context, CL_MEM_READ_ONLY, sizeof(bitCapInt));
 
-    queue.enqueueMapBuffer(stateBuffer, CL_TRUE, CL_MAP_READ | CL_MAP_WRITE, 0, sizeof(complex) * maxQPower);
+    queue.enqueueMapBuffer(*stateBuffer, CL_TRUE, CL_MAP_READ | CL_MAP_WRITE, 0, sizeof(complex) * maxQPower);
 }
 
 void QEngineOCL::ReInitOCL()
 {
     clObj = OCLEngine::Instance();
 
-    queue = *(clObj->GetQueuePtr());
-    cl::Context context = *(clObj->GetContextPtr());
-
     // create buffers on device (allocate space on GPU)
-    stateBuffer = cl::Buffer(context, CL_MEM_USE_HOST_PTR | CL_MEM_READ_ONLY, sizeof(complex) * maxQPower, stateVec);
+    stateBuffer = std::make_shared<cl::Buffer>(
+        context, CL_MEM_USE_HOST_PTR | CL_MEM_READ_ONLY, sizeof(complex) * maxQPower, stateVec);
 
-    queue.enqueueMapBuffer(stateBuffer, CL_TRUE, CL_MAP_READ | CL_MAP_WRITE, 0, sizeof(complex) * maxQPower);
+    queue.enqueueMapBuffer(*stateBuffer, CL_TRUE, CL_MAP_READ | CL_MAP_WRITE, 0, sizeof(complex) * maxQPower);
 }
 
 void QEngineOCL::ResetStateVec(complex* nStateVec)
 {
-    queue.enqueueUnmapMemObject(stateBuffer, stateVec);
+    LockGuard locked_call(device_context->mutex);
+    queue.enqueueUnmapMemObject(*stateBuffer, stateVec);
     QEngineCPU::ResetStateVec(nStateVec);
     ReInitOCL();
```

Comment:
Yes, it's both necessary and the cheaper option.

We copy in many API calls from an input state vector to an output state vector, to avoid the major complication of swapping all elements in place. To do this with OCL, we have to (first unmap the original buffer,) allocate a new buffer, then run a kernel that takes the input buffer to the output buffer. So, we already have a new buffer. The cheapest way is to have it directly replace the old buffer.

The reason for the map, is that it's not safe for anything but a kernel to touch an unmapped host pointer buffer, and it's not safe for kernels to touch mapped buffers. We want every call to complete in a state where host manipulation of the stateVec is valid. To do that, we have to ensure that it's mapped at the end. To make a kernel call, we have to first unmap. This way, we always know exactly what's going on: insofar as we rely on inheritance from QEngineCPU, we have to b
guarantee that a call always starts with a mapped buffer. QEngineOCL calls always know they need to start by unmapping the buffer, then making a kernel call, then mapping the buffer before returning.

This doesn't add the overhead that you might think it does. Mapping the buffer with a blocking call mostly just waits for the kernel to actually finish.